Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -108,6 +140,52 @@ public struct BSONDocumentIterator: IteratorProtocol {
             }
         }
     }
+
+    // uses an iterator to copy (key, value) pairs of the provided document from range [startIndex, endIndex) into a new
+    // document. starts at the startIndex-th pair and ends at the end of the document or the (endIndex-1)th index,
+    // whichever comes first.
+    internal static func subsequence(
```

Comment:
optimizing for performance is a non-goal of this project, so I think the strategy for implementing this is fine as of now. That said, we'll want to improve this in the future to just slice up the requisite bytes into a new `BSONDocument`, rather than doing 2 iterations over the document. I filed [SWIFT-911](https://jira.mongodb.org/browse/SWIFT-911) for the work covering that.

In the future, if we amend `BSONDocument` to contain further metadata about byte ranges, we could implement this without any walks over the document at all.